google/cloud/pubsub/samples/iam_samples.cc

```diff
@@ -0,0 +1,319 @@
+              << response->DebugString() << "\n";
+  }
+  // [END pubsub_get_subscription_policy]
+  (argv.at(0), argv.at(1));
+}
+
+void SetSubscriptionPolicy(std::vector<std::string> const& argv) {
+  // [START pubsub_set_subscription_policy]
+  namespace iam = google::cloud::iam;
+  namespace pubsub = google::cloud::pubsub;
+  using google::cloud::StatusCode;
+  [](std::string project_id, std::string subscription_id) {
+    auto const subscription =
+        pubsub::Subscription(std::move(project_id), std::move(subscription_id));
+    auto client = iam::IAMPolicyClient(
+        iam::MakeIAMPolicyConnection(pubsub::IAMPolicyOptions()));
+
+    // In production code, consider an OCC loop to handle concurrent changes
+    // to the policy.
+    google::iam::v1::GetIamPolicyRequest get;
+    get.set_resource(subscription.FullName());
+    auto policy = client.GetIamPolicy(get);
+    if (!policy) throw std::runtime_error(policy.status().message());
+
+    google::iam::v1::SetIamPolicyRequest set;
+    set.set_resource(subscription.FullName());
+    *set.mutable_policy() = *std::move(policy);
+    // Add all users as viewers.
+    auto& b0 = *set.mutable_policy()->add_bindings();
+    b0.set_role("roles/pubsub.viewer");
+    b0.add_members("domain:google.com");
+    // Add a group as an editor.
+    auto& b1 = *set.mutable_policy()->add_bindings();
+    b1.set_role("roles/editor");
+    b1.add_members("group:cloud-logs@google.com");
+
+    auto response = client.SetIamPolicy(set);
+    if (!response) throw std::runtime_error(response.status().message());
+    std::cout << "Policy for subscription " << subscription.FullName() << ": "
+              << response->DebugString() << "\n";
+  }
+  // [END pubsub_set_subscription_policy]
+  (argv.at(0), argv.at(1));
+}
+
+void TestSubscriptionPermissions(std::vector<std::string> const& argv) {
+  // [START pubsub_test_subscription_permissions]
+  namespace iam = google::cloud::iam;
+  namespace pubsub = google::cloud::pubsub;
+  [](std::string project_id, std::string subscription_id) {
+    auto const subscription =
+        pubsub::Subscription(std::move(project_id), std::move(subscription_id));
+    auto client = iam::IAMPolicyClient(
+        iam::MakeIAMPolicyConnection(pubsub::IAMPolicyOptions()));
+    google::iam::v1::TestIamPermissionsRequest request;
+    request.set_resource(subscription.FullName());
+    request.add_permissions("pubsub.subscriptions.consume");
+    request.add_permissions("pubsub.subscriptions.update");
+    std::cout << "request=" << request.DebugString() << "\n";
+
+    auto response = client.TestIamPermissions(request);
+    if (!response) throw std::runtime_error(response.status().message());
+    std::cout << "Allowed permissions for subscription "
+              << subscription.FullName() << ":";
+    for (auto const& permission : response->permissions()) {
+      std::cout << " " << permission;
+    }
+    std::cout << "\n";
+  }
+  // [END pubsub_test_topic_permissions]
```

Comment:
also, the bot is mad about this one. s/topic/subscription/

Reply:
Fixed